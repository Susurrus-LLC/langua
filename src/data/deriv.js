export const defData = {
  words: 5,
  type: 'any',
  lexemes: [
    {
      lexeme: 'nzi',
      definition: 'tree'
    },
    {
      lexeme: 'ilu',
      definition: 'flower'
    },
    {
      lexeme: 'ru',
      definition: 'person'
    },
    {
      lexeme: 'kwərə',
      definition: 'goat'
    },
    {
      lexeme: 'lɛtsi',
      definition: 'sword'
    }
  ],
  derivations: [
    {
      derivation: '-vo',
      gloss: 'person',
      definition:
        'Creates a noun referring to a person associated with the base.'
    },
    {
      derivation: '-lu',
      gloss: 'place',
      definition: 'Creates a noun denoting the location of the base.'
    },
    {
      derivation: 'ña-',
      gloss: 'abstract',
      definition: 'Creates an abstract noun.'
    },
    {
      derivation: 'k-sa',
      gloss: 'collective',
      definition: 'Creates a noun referring to a collection of bases.'
    }
  ],
  results: undefined
}
