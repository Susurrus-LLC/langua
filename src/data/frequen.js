export const defData = {
  corpus:
    'Nzɔ de’áxa Tsənaxa, ño lo sɛtsa’ɔk, ño lo de’áxa u Kasaga, ño lo mɛya ayru Tol o Əbe kwə, sa nzɔ xa mpaskale: Steyə rabɛ lu ñire lu mvunə poxe o lu pɛlɔ ah lo mɛwɛ axe, maxə ɔra latsi əbɔwə səyo.',
  consonants:
    'b/d/f/g/k/kw/l/m/mp/mv/n/nt/nts/nz/ñ/ŋk/ŋkw/p/r/s/t/ts/v/w/x,h/y/’',
  vowels: 'a,á/ɛ,ɛ́/e,é/ə,ə́/i,í/ɔ,ɔ́/o,ó/u,ú',
  distinguishCase: false,
  analyzed: false,
  results: undefined,
  hovered: undefined,
  filterSeg: 'all',
  filterSyll: 'all',
  filterWords: 'all'
}
