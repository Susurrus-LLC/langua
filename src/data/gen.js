export const defData = {
  subpatterns: [
    {
      selected: 'V',
      subpattern: 'a/e/i/o/u'
    },
    {
      selected: 'C',
      subpattern: 'p/t/k/b/d/g/s/m/n/l/r'
    },
    {
      selected: 'N',
      subpattern: 'm/n'
    }
  ],
  pattern: 'V/CV/CVN/CVCV/CVCVN',
  words: 100,
  newline: false,
  filterdupes: false
}

export const vars = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z'
]
