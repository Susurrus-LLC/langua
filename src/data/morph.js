export const defData = {
  categories: [
    'V=ieaou',
    'L=īēāōū',
    'C=ptcqbdgmnlrhs',
    'F=ie',
    'B=ou',
    'S=ptc',
    'Z=bdg'
  ],
  rewriteRules: ['lh=lj'],
  soundChanges: [
    '[sm]//_#',
    'i/j/_V',
    'L/V/_',
    'e//Vr_#',
    'v//V_V',
    'u/o/_#',
    'gn/nh/_',
    'S/Z/V_V',
    'c/i/F_t',
    'c/u/B_t',
    'p//V_t',
    'ii/i/_',
    'e//C_rV'
  ],
  lexicon: [
    'lector',
    'doctor',
    'focus',
    'jocus',
    'districtus',
    'cīvitatem',
    'adoptare',
    'opera',
    'secundus',
    'fīliam',
    'pōntem'
  ],
  outputFormat: 'oo',
  showDiff: true,
  showChanges: true,
  ruleReport: true,
  rewriteOutput: true,
  results: undefined
}
