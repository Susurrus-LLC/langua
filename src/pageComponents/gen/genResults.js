import React from 'react'
import PropTypes from 'prop-types'

import Results from '../../components/results'

import sharedStyles from '../../components/results/sharedResults.module.sass'

const GenResults = () => <Results />

GenResults.propTypes = {
  styles: PropTypes.string
}

export default GenResults
