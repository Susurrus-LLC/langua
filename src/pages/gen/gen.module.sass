@import ../../styles/variables.module

.duplicate
  outline: 1px solid $red

.subps-row, .adding-row, .pattern-input
  width: 100%

.subps-row, .adding-row
  display: flex
  flex-direction: row
  align-items: baseline
  align-content: space-between

.subps-row
  justify-content: space-between

.adding-row, .pattern-input
  margin: 1px 0
  padding: 1px

.adding-row
  justify-content: flex-end

.subpattern-input
  flex-grow: 1

.words-input
  width: $ms6 * 1rem

.out-text
  white-space: pre-wrap
