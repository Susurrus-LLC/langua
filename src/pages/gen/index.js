import React, { useState } from 'react'
import { graphql, useStaticQuery } from 'gatsby'

import Tool from '../../components/tool'

import GenForm from '../../pageComponents/gen/genForm'
import GenResults from '../../pageComponents/gen/genResults'
import genService from '../../services/genService'

import styles from './gen.module.sass'

export const PureGen = ({ data }) => {
  const toolInfo = data?.site?.siteMetadata?.toolInfo?.gen

  // State

  const [subpatterns, setSubpatterns] = useState(
    genService.getData().subpatterns
  )
  const [pattern, setPattern] = useState(genService.getData().pattern)
  const [words, setWords] = useState(genService.getData().words)
  const [newline, setNewline] = useState(genService.getData().newline)
  const [filterdupes, setFilterdupes] = useState(
    genService.getData().filterdupes
  )
  const [status, setStatus] = useState(['ok'])
  const [results, setResults] = useState([''])
  const [stats, setStats] = useState({
    words: 0,
    maxWords: 0,
    filtered: 0,
    remaining: 0
  })

  // Functions

  const onChangeInput = e => {
    const data = {
      subpatterns,
      pattern,
      words,
      newline,
      filterdupes
    }

    if (e.target.tagName === 'BUTTON') {
      e.preventDefault()
      if (e.target.type === 'submit') {
        // Generate the output
        const response = genService.generate(data)
        setStatus(response.status)
        setResults(response.results)
        setStats(response.stats)
        // Save the current state to storage
        genService.setStorage(data)
      } else if (e.target.type === 'button') {
        if (e.target.id === 'save') {
          // Save the current state to storage and generate a file
          genService.save(data)
        } else if (e.target.id === 'add') {
          // When the add button is clicked, add a blank Subpattern to state
          setSubpatterns(genService.add(data).subpatterns)
        } else if (e.target.id.slice(0, 1) === 'c') {
          // When a Subpattern is cleared, delete the corresponding Subpattern from state
          const id = e.target.id.slice(1)
          setSubpatterns(genService.clear(id, data).subpatterns)
        }
      }
    } else if (e.target.type === 'file') {
      e.preventDefault()
      // Open a file and parse it to restore a saved state
      const file = e.target.files[0]
      const updateState = response => {
        if (response) {
          // Only change state if the file was successfully opened
          setSubpatterns(response.subpatterns)
          setPattern(response.pattern)
          setWords(response.words)
          setNewline(response.newline)
          setFilterdupes(response.filterdupes)
        }
      }
      genService.open(file, updateState)
    } else if (e.target.type === 'checkbox') {
      const checked = e.target.checked
      if (e.target.id === 'newline') {
        // If the selection for new lines is changed, store that change in state
        setNewline(genService.changeNewline(checked, data).newline)
      } else if (e.target.id === 'filterdupes') {
        // If the selection for filtering duplicates is changed, store that change in state
        setFilterdupes(genService.changeDupes(checked, data).filterdupes)
      }
    } else {
      const val = e.target.value
      if (e.target.id === 'pattern') {
        // When the pattern is changed, store the change in state
        setPattern(genService.changePattern(val, data).pattern)
      } else if (e.target.id === 'words') {
        // When the number of desired words is changed, store that change in state
        const response = genService.wordNumChange(val, data)
        if (response) {
          // Only update state if there's a change
          setWords(response.words)
        }
      } else {
        const id = e.target.id.slice(1)
        const which = e.target.id.slice(0, 1)
        if (which === 'v') {
          // When a Subpattern variable is changed, store that change in state
          setSubpatterns(genService.changeSelect(id, val, data).subpatterns)
        } else if (which === 'p') {
          // When a Subpattern is changed, store that change in state
          setSubpatterns(genService.changeSubpattern(id, val, data).subpatterns)
        }
      }
    }
  }

  return (
    <Tool toolInfo={toolInfo}>
      <GenForm
        styles={styles}
        data={{
          subpatterns,
          pattern,
          words,
          newline,
          filterdupes
        }}
        change={onChangeInput}
      />
      <GenResults
        styles={styles}
        newline={newline}
        filterDupes={filterdupes}
        status={status}
        results={results}
        stats={stats}
      />
    </Tool>
  )
}

const Gen = props => {
  const data = useStaticQuery(graphql`
    query GenToolQuery {
      site {
        siteMetadata {
          toolInfo {
            gen {
              title
              link
              description
            }
          }
        }
      }
    }
  `)

  return <PureGen {...props} data={data} />
}

export default Gen
