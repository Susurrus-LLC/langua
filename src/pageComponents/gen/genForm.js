import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'

import Form from '../../components/form'
import Button from '../../components/button'
import Control from '../../components/control'
import ControlSide from '../../componnets/control/side'
import ControlPiece from '../../components/control/piece'

import sharedStyles from '../../components/form/sharedForm.module.sass'

import { vars } from '../../data/gen'

const GenForm = () => {
  const filePicker = React.createRef()

  const selectOptions = vars.map((variab, index) => (
    <option key={index} value={variab}>
      {variab}
    </option>
  ))

  const subpatternRows = data.subpatterns.map((subpattern, index) => (
    <div key={index} className={styles.subpsRow}>
      <select
        id={`v${index}`}
        name={`v${index}`}
        value={subpattern.selected}
        onChange={change}
        className={styles.select}
        aria-label={`Variable name for subpattern ${index}`}
      >
        {selectOptions}
      </select>
      <input
        type='text'
        id={`p${index}`}
        name={`p${index}`}
        className={styles.subpatternInput}
        value={subpattern.subpattern}
        onChange={change}
        aria-label={`Subpattern ${index}`}
      />
      <div className={styles.clearButton}>
        <Button
          id={`c${index}`}
          onCLick={change}
          ver='danger'
          small
          type='button'
          aria-label={`Delete subpattern ${index}`}
        >
          Delete
        </Button>
      </div>
    </div>
  ))

  return <Form name='gen-form' />
}

GenForm.propTypes = {
  styles: PropTypes.string
}

export default GenForm
