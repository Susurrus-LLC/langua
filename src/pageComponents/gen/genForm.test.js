import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'

import GenForm from './genForm'
import { defData } from '../../data/gen'

describe('GenForm', () => {
  const styles = {
    subpsRow: 'subpsRow',
    select: 'select',
    subpatternInput: 'subpatternInput',
    clearButton: 'clearButton',
    addingRow: 'addingRow',
    subpatterns: 'subpatterns',
    pattern: 'pattern',
    patternInput: 'patternInput',
    wordsInput: 'wordsInput'
  }

  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <GenForm styles={styles} data={defData} change={jest.fn()} />,
      div
    )
    ReactDOM.unmountComponentAtNode(div)
  })

  it('renders accurately', () => {
    let tree = renderer
      .create(<GenForm styles={styles} data={defData} change={jest.fn()} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
