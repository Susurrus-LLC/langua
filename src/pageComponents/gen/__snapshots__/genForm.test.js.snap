// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenForm renders accurately 1`] = `
<form
  className="form"
  name="gen-form"
>
  <div
    className="subpatterns part"
  >
    <h5
      className="sectionTitle"
    >
      Subpatterns
    </h5>
    <div
      className="subpsRow"
    >
      <select
        aria-label="Variable name for subpattern 0"
        className="select"
        id="v0"
        name="v0"
        onChange={[MockFunction]}
        value="V"
      >
        <option
          value="A"
        >
          A
        </option>
        <option
          value="B"
        >
          B
        </option>
        <option
          value="C"
        >
          C
        </option>
        <option
          value="D"
        >
          D
        </option>
        <option
          value="E"
        >
          E
        </option>
        <option
          value="F"
        >
          F
        </option>
        <option
          value="G"
        >
          G
        </option>
        <option
          value="H"
        >
          H
        </option>
        <option
          value="I"
        >
          I
        </option>
        <option
          value="J"
        >
          J
        </option>
        <option
          value="K"
        >
          K
        </option>
        <option
          value="L"
        >
          L
        </option>
        <option
          value="M"
        >
          M
        </option>
        <option
          value="N"
        >
          N
        </option>
        <option
          value="O"
        >
          O
        </option>
        <option
          value="P"
        >
          P
        </option>
        <option
          value="Q"
        >
          Q
        </option>
        <option
          value="R"
        >
          R
        </option>
        <option
          value="S"
        >
          S
        </option>
        <option
          value="T"
        >
          T
        </option>
        <option
          value="U"
        >
          U
        </option>
        <option
          value="V"
        >
          V
        </option>
        <option
          value="W"
        >
          W
        </option>
        <option
          value="X"
        >
          X
        </option>
        <option
          value="Y"
        >
          Y
        </option>
        <option
          value="Z"
        >
          Z
        </option>
      </select>
      <input
        aria-label="Subpattern 0"
        className="subpatternInput"
        id="p0"
        name="p0"
        onChange={[MockFunction]}
        type="text"
        value="a/e/i/o/u"
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="c0"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="subpsRow"
    >
      <select
        aria-label="Variable name for subpattern 1"
        className="select"
        id="v1"
        name="v1"
        onChange={[MockFunction]}
        value="C"
      >
        <option
          value="A"
        >
          A
        </option>
        <option
          value="B"
        >
          B
        </option>
        <option
          value="C"
        >
          C
        </option>
        <option
          value="D"
        >
          D
        </option>
        <option
          value="E"
        >
          E
        </option>
        <option
          value="F"
        >
          F
        </option>
        <option
          value="G"
        >
          G
        </option>
        <option
          value="H"
        >
          H
        </option>
        <option
          value="I"
        >
          I
        </option>
        <option
          value="J"
        >
          J
        </option>
        <option
          value="K"
        >
          K
        </option>
        <option
          value="L"
        >
          L
        </option>
        <option
          value="M"
        >
          M
        </option>
        <option
          value="N"
        >
          N
        </option>
        <option
          value="O"
        >
          O
        </option>
        <option
          value="P"
        >
          P
        </option>
        <option
          value="Q"
        >
          Q
        </option>
        <option
          value="R"
        >
          R
        </option>
        <option
          value="S"
        >
          S
        </option>
        <option
          value="T"
        >
          T
        </option>
        <option
          value="U"
        >
          U
        </option>
        <option
          value="V"
        >
          V
        </option>
        <option
          value="W"
        >
          W
        </option>
        <option
          value="X"
        >
          X
        </option>
        <option
          value="Y"
        >
          Y
        </option>
        <option
          value="Z"
        >
          Z
        </option>
      </select>
      <input
        aria-label="Subpattern 1"
        className="subpatternInput"
        id="p1"
        name="p1"
        onChange={[MockFunction]}
        type="text"
        value="p/t/k/b/d/g/s/m/n/l/r"
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="c1"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="subpsRow"
    >
      <select
        aria-label="Variable name for subpattern 2"
        className="select"
        id="v2"
        name="v2"
        onChange={[MockFunction]}
        value="N"
      >
        <option
          value="A"
        >
          A
        </option>
        <option
          value="B"
        >
          B
        </option>
        <option
          value="C"
        >
          C
        </option>
        <option
          value="D"
        >
          D
        </option>
        <option
          value="E"
        >
          E
        </option>
        <option
          value="F"
        >
          F
        </option>
        <option
          value="G"
        >
          G
        </option>
        <option
          value="H"
        >
          H
        </option>
        <option
          value="I"
        >
          I
        </option>
        <option
          value="J"
        >
          J
        </option>
        <option
          value="K"
        >
          K
        </option>
        <option
          value="L"
        >
          L
        </option>
        <option
          value="M"
        >
          M
        </option>
        <option
          value="N"
        >
          N
        </option>
        <option
          value="O"
        >
          O
        </option>
        <option
          value="P"
        >
          P
        </option>
        <option
          value="Q"
        >
          Q
        </option>
        <option
          value="R"
        >
          R
        </option>
        <option
          value="S"
        >
          S
        </option>
        <option
          value="T"
        >
          T
        </option>
        <option
          value="U"
        >
          U
        </option>
        <option
          value="V"
        >
          V
        </option>
        <option
          value="W"
        >
          W
        </option>
        <option
          value="X"
        >
          X
        </option>
        <option
          value="Y"
        >
          Y
        </option>
        <option
          value="Z"
        >
          Z
        </option>
      </select>
      <input
        aria-label="Subpattern 2"
        className="subpatternInput"
        id="p2"
        name="p2"
        onChange={[MockFunction]}
        type="text"
        value="m/n"
      />
      <div
        className="clearButton"
      >
        <button
          className="btn danger small"
          id="c2"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
    <div
      className="addingRow"
    >
      <button
        className="btn success small"
        id="add"
        onClick={[MockFunction]}
        role="button"
        type="button"
      >
        Add
      </button>
    </div>
  </div>
  <div
    className="pattern part"
  >
    <h5
      className="sectionTitle"
    >
      Pattern
    </h5>
    <input
      aria-label="Pattern"
      className="patternInput"
      id="pattern"
      name="pattern"
      onChange={[MockFunction]}
      type="text"
      value="V/CV/CVN/CVCV/CVCVN"
    />
  </div>
  <div
    className="control part"
  >
    <div
      className="controlSide controlLeft"
    >
      <div
        className="controlPiece"
      >
        <button
          className="btn neutral"
          id="generate"
          onClick={[MockFunction]}
          role="button"
          type="submit"
        >
          Generate
        </button>
      </div>
      <div
        className="controlPiece"
      >
        <label
          htmlFor="words"
        >
          words:
        </label>
        <input
          aria-label="Number of words to generate"
          className="wordsInput"
          id="words"
          max="9999"
          min="1"
          name="words"
          onChange={[MockFunction]}
          type="number"
          value={100}
        />
      </div>
      <div
        className="controlPiece"
      >
        <label>
          <input
            checked={false}
            id="newline"
            name="options"
            onChange={[MockFunction]}
            type="checkbox"
            value="newline"
          />
           
          new line each
        </label>
      </div>
      <div
        className="controlPiece"
      >
        <label>
          <input
            checked={false}
            id="filterdupes"
            name="options"
            onChange={[MockFunction]}
            type="checkbox"
            value="filterdupes"
          />
           
          filter duplicates
        </label>
      </div>
    </div>
    <div
      className="controlSide controlRight"
    >
      <div
        className="controlPiece"
      >
        <button
          className="btn success"
          id="save"
          onClick={[MockFunction]}
          role="button"
          type="button"
        >
          Save
        </button>
      </div>
      <div
        className="controlPiece"
      >
        <button
          className="btn"
          id="open"
          onClick={[Function]}
          role="button"
          type="button"
        >
          Open
        </button>
        <input
          className="hidden"
          id="file"
          name="file"
          onChange={[MockFunction]}
          type="file"
        />
      </div>
    </div>
  </div>
</form>
`;
